runtime/parachains/src/scheduler.rs

```diff
@@ -680,83 +382,212 @@ impl<T: Config> Pallet<T> {
+					CoreIndex::from(core_idx as u32),
+					ass,
+				),
+			}
 		}
 	}
 
-	// Free all scheduled cores and return parathread claims to queue, with retries incremented.
-	pub(crate) fn clear() {
-		let config = <configuration::Pallet<T>>::config();
-		ParathreadQueue::<T>::mutate(|queue| {
-			for core_assignment in Scheduled::<T>::take() {
-				if let AssignmentKind::Parathread(collator, retries) = core_assignment.kind {
-					if !<paras::Pallet<T>>::is_parathread(core_assignment.para_id) {
-						continue
-					}
+	//
+	//  ClaimQueue related functions
+	//
+	fn claimqueue_lookahead() -> u32 {
+		match <configuration::Pallet<T>>::config().scheduling_lookahead {
+			0 => 1,
+			n => n,
+		}
+	}
 
-					let entry = ParathreadEntry {
-						claim: ParathreadClaim(core_assignment.para_id, collator),
-						retries: retries + 1,
-					};
+	// on new session
+	fn clear_claimqueue() {
+		for (core_idx, cqv) in ClaimQueue::<T>::take() {
+			for ca in cqv.into_iter().flatten() {
+				T::AssignmentProvider::push_assignment_for_core(
+					core_idx,
+					Assignment::from_core_assignment(ca),
+				);
+			}
+		}
+	}
+
+	pub(crate) fn update_claimqueue(
+		just_freed_cores: BTreeMap<CoreIndex, FreedReason>,
+		now: T::BlockNumber,
+	) -> Vec<CoreAssignment> {
+		Self::move_claimqueue_forward();
+		Self::fill_claimqueue(just_freed_cores, now)
+	}
 
-					if entry.retries <= config.parathread_retries {
-						queue.enqueue_entry(entry, config.parathread_cores);
+	fn move_claimqueue_forward() {
+		let mut cq = ClaimQueue::<T>::get();
+		for (_, vec) in cq.iter_mut() {
+			match vec.front() {
+				None => {},
+				Some(None) => {
+					vec.pop_front();
+				},
+				Some(_) => {},
+			}
+		}
+
+		ClaimQueue::<T>::set(cq);
+	}
+
+	fn fill_claimqueue(
+		just_freed_cores: BTreeMap<CoreIndex, FreedReason>,
+		now: T::BlockNumber,
+	) -> Vec<CoreAssignment> {
+		let (mut concluded_paras, mut timedout_paras) = Self::free_cores(just_freed_cores);
+
+		// This can only happen on new sessions at which we move all assignments back to the provider.
+		// Hence, there's nothing we need to do here.
+		if ValidatorGroups::<T>::get().is_empty() {
+			vec![]
+		} else {
+			let n_lookahead = Self::claimqueue_lookahead();
+			let n_session_cores = T::AssignmentProvider::session_core_count();
+			let cq = ClaimQueue::<T>::get();
+
+			for core_idx in 0..n_session_cores {
+				let core_idx = CoreIndex(core_idx);
+				let group_idx = Self::group_assigned_to_core(core_idx, now).expect(
+					"core is not out of bounds and we are guaranteed \
+										  to be after the most recent session start; qed",
+				);
+
+				// add previously timedout paras back into the queue
+				if let Some(assignment) = timedout_paras.remove(&core_idx) {
+					let ca = assignment.to_core_assignment(core_idx, group_idx);
+					Self::add_to_claimqueue(ca);
+				}
+
+				let n_lookahead_used = cq.get(&core_idx).map_or(0, |v| v.len() as u32);
```

Comment:
> I thought about that. Not sure that is a problem in general as time outs are expected to be rare?

If we punt on an edge case we should the least have a comment describing the edge case and arguing why it is fine to not bother. Better to not have the edge case in the first place though. I usually only do that if it either adds lots of complexity, has a big performance hit or both (and the edge case does no real harm of course).

> Good question! It's what the old code did. I assume that because the number of cores could have changed on a new session?

I guess that should be handled on session change?

Reply:
> If we punt on an edge case we should the least have a comment describing the edge case and arguing why it is fine to not bother. Better to not have the edge case in the first place though. I usually only do that if it either adds lots of complexity, has a big performance hit or both (and the edge case does no real harm of course).

Alright, I'll look into making sure the claimqueue doesn't go over the desired size.
> I guess that should be handled on session change?

Sorry, I was talking about the claimqueue being filled for that cores. The number of cores can change on session change but they don't get filled there. Which I suppose we could do and then only iterate over freed cores here.

Reply:
> Alright, I'll look into making sure the claimqueue doesn't go over the desired size.

Since this only happens with timedout paras that should be retried, should we push back them to the assignment provider here?

Reply:
No we talked about this. Pushing back on retry adds unnecessary complexity as then the assignment provider needs to track the number of already happened retries. Ok technically this is easily possible now as you changed the `AssignmentProvider` to handle `ParasEntry`s which contain the number of retries. 

Which makes me realize: renaming to `ParasEntry` is totally fine (commented on the other PR), but `ParasEntry` should be local to the scheduler. What is coming from the assignment provider should still be `Assignment`s. This way the scheduler can put into `ParasEntry` whatever it wants/needs and the assignment providers are not affected. Given that retries are already part of the interface it is debatable whether having `retries` exposed to the `AssignmentProvider` is that big a deal, but they absolutely don't care about this value, they don't even need to know what it is, yet they still have to provide some value.

I think the fix is easy: Just consider an occupied core as part of the claim queue, in other words:

```rust
				let n_lookahead_used = cq.get(&core_idx).map_or(0, |v| v.len() as u32) + if is_core_occupied(core_idx) { 1 } else { 0};
```